pkg/services/service.go

```diff
@@ -0,0 +1,232 @@
+package services
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/exp/maps"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+)
+
+// Service represents a long-running service inside the Application.
+//
+// The simplest way to implement a Service is via NewService.
+//
+// For other cases, consider embedding a services.StateMachine to implement these
+// calls in a safe manner.
+type Service interface {
+	// Start the service.
+	//  - Must return promptly if the context is cancelled.
+	//  - Must not retain the context after returning (only applies to start-up)
+	//  - Must not depend on external resources (no blocking network calls)
+	Start(context.Context) error
+	// Close stops the Service.
+	// Invariants: Usually after this call the Service cannot be started
+	// again, you need to build a new Service to do so.
+	//
+	// See MultiCloser
+	Close() error
+
+	HealthReporter
+}
+
+// Engine manages service internals like health, goroutine tracking, and shutdown signals.
+type Engine struct {
+	StopChan
+	logger.SugaredLogger
+
+	wg sync.WaitGroup
+
+	healthEvent func(error)
+	conds       map[string]error
+	condsMu     sync.RWMutex
+}
+
+// Go calls fn in a tracked goroutine that will block closing the service.
+// fn should yield to StopChan promptly.
+func (e *Engine) Go(fn func()) {
+	e.wg.Add(1)
+	go func() {
+		defer e.wg.Done()
+		fn()
+	}()
+}
+
+func (e *Engine) GoRepeat(period time.Duration, fn func()) {
+	e.Go(func() {
+		for {
+			select {
+			case <-e.StopChan:
+				return
+			case <-time.After(period): //TODO jitter?
+				fn()
+			}
+		}
+	})
+}
+
+// HealthEvent records an error to be reported via the next call to Healthy().
+func (e *Engine) HealthEvent(err error) { e.healthEvent(err) }
+
+// SetUnhealthy records a condition key and an error, which causes an unhealthy report, until SetHealthy(condition) is called.
+func (e *Engine) SetUnhealthy(condition string, err error) {
+	e.condsMu.Lock()
+	defer e.condsMu.Unlock()
+	e.conds[condition] = fmt.Errorf("%s: %e", condition, err)
+}
+
+// SetHealthy removes a condition and error recorded by SetUnhealthy.
+func (e *Engine) SetHealthy(condition string) {
+	e.condsMu.Lock()
+	defer e.condsMu.Unlock()
+	delete(e.conds, condition)
+}
+
+// Unhealthy causes an unhealthy report, until the returned func() is called.
+// Use this for simple cases where the func() can be kept in scope, and prefer to defer it inline if possible:
+//
+//	defer Unhealthy(fmt.Errorf("foo bar: %i", err))()
+//
+// See SetUnhealthy for an alternative API.
+func (e *Engine) Unhealthy(err error) func() {
```

Comment:
I am still not satisfied with these names. I think that these four different operations are fundamental, but I'm not sure what to call them that would be more intuitive than this... :thinking: